File: client/src/game/utils/collision.lua
local Collision = {}

function noop()
end

function Collision.check(a, b, checks)
  local check1 = checks[1] or noop
  local check2 = checks[2] or noop

  if check1(a) and check2(b) then
    return {Collision.entity(a), Collision.entity(b)}
  elseif check1(b) and check2(a) then
    return {Collision.entity(b), Collision.entity(a)}
  end

  return false
end

function Collision.entity(fixture)
  return fixture:getUserData().entity
end

function Collision.is(type, fixture)
  return Collision.entity(fixture).meta.type == type
end

function Collision.hasComponent(component)
  return function(fixture)
    return Collision.entity(fixture):has(component)
  end
end

function Collision.as(component, fixture)
  return Collision.entity(fixture):as(component)
end

function Collision.isNotDynamic(fixture)
  return fixture:getBody():getType() ~= 'dynamic'
end

function Collision.isBodyType(type)
  return function(fixture)
    return fixture:getBody():getType() == type
  end
end

function Collision.isNotBodyType(type)
  return function(fixture)
    return fixture:getBody():getType() ~= type
  end
end

function Collision.isAggression(fixture)
  return fixture:getUserData().aggression
end

return Collision

File: client2/src/game/utils/collision.ts
import { Component } from '../../ecs/component'
import { Entity } from '../../ecs/Entity'

const noop = () => {}

export const getEntity = (fixture: Fixture): Entity => {
  return fixture.getUserData().entity
}

export const hasComponent = (component: Component) => (fixture: Fixture) => {
  return getEntity(fixture).has(component)
}

export const hasComponents = (...components: Component[]) => (
  fixture: Fixture
) => {
  return components.every(component => hasComponent(component)(fixture))
}

export const isBodyType = (type: BodyType) => (fixture: Fixture) => {
  return fixture.getBody().getType() === type
}

export const isNotBodyType = (type: BodyType) => (fixture: Fixture) => {
  return !isBodyType(type)(fixture)
}
export const isSensor = (fixture: Fixture) => {
  return fixture.isSensor()
}

export const isNotSensor = (fixture: Fixture) => {
  return !isSensor(fixture)
}

export const check = (
  a: Fixture,
  b: Fixture,
  checks: [(fixture: Fixture) => boolean, (fixture: Fixture) => boolean]
) => {
  const check1 = checks[0] || noop
  const check2 = checks[1] || noop

  if (check1(a) && check2(b)) {
    return [getEntity(a), getEntity(b)]
  } else if (check1(b) && check2(a)) {
    return [getEntity(b), getEntity(a)]
  }

  return null
}
